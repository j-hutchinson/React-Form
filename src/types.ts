
export enum FormState {
    USER = 'User',
    PRIVACY = 'Privacy',
    DONE = 'Done',
}

export interface ReduxState {
    user: UserSettings;
    privacy: PrivacySettings;
}

export interface UserSettings {
    name: string;
    title: string;
    email: string;
    password: string
}

export interface PrivacySettings {
    updates: boolean;
    communication: boolean;
}

export interface ActionType {
    type: string;
    privacy?: PrivacySettings;
    user?: UserSettings;
}