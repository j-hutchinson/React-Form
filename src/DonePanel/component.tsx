import React from 'react';
import styled from 'styled-components';
import { PanelStyles } from '../css/tokens'
import store from '../store/store'

const StyledPanel = styled.div`
    ${PanelStyles}
`;

const DonePanel = (): JSX.Element => {
    console.log(store.getState())
    return (
        <StyledPanel>Please verify your email address, you should have received an email from us today</StyledPanel>
    );
}

export default DonePanel;