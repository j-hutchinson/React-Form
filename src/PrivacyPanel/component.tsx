import React, { useState } from 'react';
import styled from 'styled-components';
import { PanelStyles } from '../css/tokens'
import store from '../store/store'
import { updatePrivacy } from '../store/actions/actions';

interface Props {
    onDoneClick: () => void;
}

export const StyledPanel = styled.form`
    ${PanelStyles};
`;

export const StyledBox = styled.div`
    height: 20px;
    width: 20px;
    border: 1px solid #B9B9B9;
    border-radius: 1px;
    padding: 4px 0 4px 2px;
    color: #737373;
    background-color: white;
    display: flex;

    :hover {
        border-color: #007398;
    }
`;

const StyledButtonWrapper = styled.div`
    display: flex;
    flex-direction: row-reverse;
    margin-right: 10px;
`;

const ConfirmIcon = () => (
    <svg width="14px" height="10px" viewBox="0 0 10 8">
        <path fill="currentColor" d="M8.8 0L10 1.1 3.4 7.7 0 4.3l1.1-1.1 2.3 2.2z" />
    </svg>
);

const PrivacyPanel = ({ onDoneClick }: Props): JSX.Element => {
    const [activeUpdates, setUpdates] = useState<boolean>(false);
    const [activeComms, setComms] = useState<boolean>(false);

    const onSubmit = () => {
        store.dispatch(updatePrivacy({ updates: activeUpdates, communication: activeComms }));
        onDoneClick();
    }

    return (
        <StyledPanel onSubmit={onSubmit}>
            <div>
                <StyledBox onClick={() => setUpdates(!activeUpdates)}>{activeUpdates && <ConfirmIcon />}</StyledBox>
                Receive updates about Tray.io product by email
            </div>
            <div>
                <StyledBox onClick={() => setComms(!activeComms)}>{activeComms && <ConfirmIcon />}</StyledBox>
                Receive communcations by email from other products created by the Tray.io team
            </div>
            <StyledButtonWrapper>
                <input type="submit" value="Submit" />
            </StyledButtonWrapper>
        </StyledPanel>
    );
}

export default PrivacyPanel;