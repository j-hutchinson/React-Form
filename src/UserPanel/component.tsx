import React, { FormEvent, useState } from 'react';
import styled from 'styled-components';
import { PanelStyles, SubmitInputStyles } from '../css/tokens';
import { nextPage, updateUser } from '../store/actions/actions';
import store from '../store/store';

export const StyledPanel = styled.form`
    ${PanelStyles}
`;

const StyledItem = styled.div`
    display: grid;
    grid-template-rows: min-content 1fr;
    grid-gap: 4px;
    margin: 10px;
`;

const StyledButtonWrapper = styled.div`
    display: flex;
    flex-direction: row-reverse;
    margin-right: 10px;
`;

const StyledAsterisk = styled.div`
    display: inline-block;
    color: red;
`;

const StyledError = styled.div`
    color: red;
    font-size: 14px;
    margin: 10px;
    font-weight: bold
`;

const StyledSubmitInput = styled.input`
   ${SubmitInputStyles};
`;

export const StyledInput = styled.input`
   height: 24px;
   font-size: 14px;
`;

/*
**  Check if password contains a digit, a lower case letter, an upper case letter, and greater than or equal to 9 characters in length
*/
export const passwordValidation = (password: string): boolean => {
    return /\d/.test(password) && /[a-z]/.test(password) && /[A-Z]/.test(password) && password.length >= 9
}

const UserPanel = (): JSX.Element => {
    const [name, setName] = useState<string>('');
    const [role, setRole] = useState<string>('');
    const [email, setEmail] = useState<string>('');
    const [password, setPassword] = useState<string>('');
    const [error, setError] = useState<boolean>(false);

    const onSubmission = (ev: FormEvent): void => {
        ev.preventDefault();
        if (passwordValidation(password)) {
            store.dispatch(updateUser({ name, role, email, password }));
            store.dispatch(nextPage());
        } else {
            setError(true);
        }
    }

    return (
        <StyledPanel onSubmit={onSubmission}>
            <StyledItem>
                <label htmlFor="name"> Name: <StyledAsterisk>*</StyledAsterisk> </label>
                <StyledInput id="name" type="text" value={name} onChange={(ev) => setName(ev.target.value)} required />
            </StyledItem>
            <StyledItem>
                <label htmlFor="role">Role:</label>
                <StyledInput id="role" type="text" value={role} onChange={(ev) => setRole(ev.target.value)} />
            </StyledItem>
            <StyledItem>
                <label htmlFor="email"> Email: <StyledAsterisk>*</StyledAsterisk> </label>
                <StyledInput id="email" type="email" value={email} onChange={(ev) => setEmail(ev.target.value)} required />
            </StyledItem>
            <StyledItem>
                <label htmlFor="password"> Password: <StyledAsterisk>*</StyledAsterisk> </label>
                <StyledInput id="password" type="password" value={password} onChange={(ev) => setPassword(ev.target.value)} required />
            </StyledItem>
            {error && (
                <StyledError>
                    Check your password is valid. Must contain 1 uppercase, 1 lowercase, 1 number and be longer than 9 characters
                </StyledError>
            )}
            <StyledButtonWrapper>
                <StyledSubmitInput type="submit" value="Submit" />
            </StyledButtonWrapper>
        </StyledPanel>
    )
};

export default UserPanel;