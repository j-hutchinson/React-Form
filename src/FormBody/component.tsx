import React from 'react';
import DonePanel from '../DonePanel/component';
import PrivacyPanel from '../PrivacyPanel/component';
import UserPanel from '../UserPanel/component';
import { FormState } from '../types';

interface Props {
    activeStep: FormState;
    setActiveStep: (key: FormState) => void;
}

const FormBody = ({ activeStep, setActiveStep }: Props): JSX.Element => (
    <div>
        {activeStep === FormState.USER && (
            <UserPanel onDoneClick={() => setActiveStep(FormState.PRIVACY)} />
        )}
        {activeStep === FormState.PRIVACY && (
            <PrivacyPanel onDoneClick={() => setActiveStep(FormState.DONE)} />
        )}
        {activeStep === FormState.DONE && <DonePanel />}
    </div>
)


export default FormBody;