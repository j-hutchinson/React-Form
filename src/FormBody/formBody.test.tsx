import { shallow } from 'enzyme';
import React from 'react';
import FormBody from './component';
import { FormState } from '../types';
import DonePanel from '../DonePanel/component';
import PrivacyPanel from '../PrivacyPanel/component';
import UserPanel from '../UserPanel/component';

describe('FormBody component', () => {
    const setActiveStep = jest.fn();

    afterEach(() => {
        jest.resetAllMocks();
    })

    test.each`
        active               | component
        ${FormState.USER}    | ${UserPanel}
        ${FormState.PRIVACY} | ${PrivacyPanel}
        ${FormState.DONE}    | ${DonePanel}
    `('component matches snapshot when $active is active', ({ active, component }) => {
        expect.assertions(2);

        const wrapper = shallow(<FormBody activeStep={active} setActiveStep={setActiveStep} />);

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(component).length).toBe(1);
    });

    test.each`
        active               | component       | nextState
        ${FormState.USER}    | ${UserPanel}    | ${FormState.PRIVACY}
        ${FormState.PRIVACY} | ${PrivacyPanel} | ${FormState.DONE}
    `('$active onDoneClick triggers next active panel', ({ active, component, nextState }) => {
        expect.assertions(2);

        const wrapper = shallow(<FormBody activeStep={active} setActiveStep={setActiveStep} />);
        wrapper.find(component).props().onDoneClick();

        expect(setActiveStep).toHaveBeenCalledTimes(1);
        expect(setActiveStep).toHaveBeenCalledWith(nextState);
    });


});
