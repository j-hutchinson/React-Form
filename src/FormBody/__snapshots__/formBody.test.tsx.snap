// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormBody component component matches snapshot when Done is active 1`] = `
<div>
  <DonePanel />
</div>
`;

exports[`FormBody component component matches snapshot when Privacy is active 1`] = `
<div>
  <PrivacyPanel
    onDoneClick={[Function]}
  />
</div>
`;

exports[`FormBody component component matches snapshot when User is active 1`] = `
<div>
  <UserPanel
    onDoneClick={[Function]}
  />
</div>
`;
